File: src/main/java/org/cubewhy/celestial/utils/GuiUtils.java
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */

package org.cubewhy.celestial.utils;

import lombok.extern.slf4j.Slf4j;
import org.cubewhy.celestial.Celestial;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import javax.swing.*;
import javax.swing.filechooser.FileFilter;
import javax.swing.filechooser.FileNameExtensionFilter;
import java.awt.*;
import java.io.File;
import java.io.IOException;

@Slf4j
public final class GuiUtils {
    private GuiUtils() {
    }

    public static @Nullable File chooseFile(FileFilter filter) {
        JFileChooser fileDialog = new JFileChooser();
        fileDialog.setFileFilter(filter);
        fileDialog.addChoosableFileFilter(filter);
        fileDialog.setFileSelectionMode(JFileChooser.FILES_ONLY);
        return (fileDialog.showOpenDialog(Celestial.launcherFrame) == JFileChooser.CANCEL_OPTION) ? null : fileDialog.getSelectedFile();
    }

    public static @Nullable File chooseFolder() {
        JFileChooser fileDialog = new JFileChooser();
        fileDialog.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY);
        return (fileDialog.showOpenDialog(Celestial.launcherFrame) == JFileChooser.CANCEL_OPTION) ? null : fileDialog.getSelectedFile();
    }

    public static File saveFile(FileNameExtensionFilter filter) {
        JFileChooser fileDialog = new JFileChooser();
        fileDialog.setFileFilter(filter);
        fileDialog.addChoosableFileFilter(filter);
        fileDialog.setFileSelectionMode(JFileChooser.FILES_ONLY);
        return (fileDialog.showSaveDialog(Celestial.launcherFrame) == JFileChooser.CANCEL_OPTION) ? null : fileDialog.getSelectedFile();
    }

    public static @NotNull JButton createButtonOpenFolder(String text, File folder) {
        JButton btn = new JButton(text);
        btn.addActionListener(e -> {
            try {
                if (folder.mkdirs()) {
                    log.info("Creating " + folder + " because the folder not exist");
                }
                Desktop.getDesktop().open(folder);
            } catch (IOException ex) {
                throw new RuntimeException(ex);
            }
        });
        return btn;
    }
}

File: src/main/kotlin/org/cubewhy/celestial/utils/GuiUtils.kt
/*
 * Celestial Launcher <me@lunarclient.top>
 * License under GPLv3
 * Do NOT remove this note if you want to copy this file.
 */
package org.cubewhy.celestial.utils

import org.cubewhy.celestial.launcherFrame
import java.awt.Desktop
import java.awt.event.ActionEvent
import java.io.File
import javax.swing.JButton
import javax.swing.JFileChooser
import javax.swing.JLabel
import javax.swing.filechooser.FileFilter
import javax.swing.filechooser.FileNameExtensionFilter


fun chooseFile(filter: FileFilter? = null): File? {
    val fileDialog = JFileChooser()
    if (filter != null) {
        fileDialog.fileFilter = filter
        fileDialog.addChoosableFileFilter(filter)
    }
    fileDialog.fileSelectionMode = JFileChooser.FILES_ONLY
    return if ((fileDialog.showOpenDialog(launcherFrame) == JFileChooser.CANCEL_OPTION)) null else fileDialog.selectedFile
}

fun chooseFolder(): File? {
    val fileDialog = JFileChooser()
    fileDialog.fileSelectionMode = JFileChooser.DIRECTORIES_ONLY
    return if ((fileDialog.showOpenDialog(launcherFrame) == JFileChooser.CANCEL_OPTION)) null else fileDialog.selectedFile
}

fun createButtonOpenFolder(text: String?, folder: File): JButton {
    val btn = JButton(text)
    btn.addActionListener {
        folder.mkdirs()
        Desktop.getDesktop().open(folder)
    }
    return btn
}


fun saveFile(filter: FileNameExtensionFilter?): File? {
    val fileDialog = JFileChooser()
    fileDialog.fileFilter = filter
    fileDialog.addChoosableFileFilter(filter)
    fileDialog.fileSelectionMode = JFileChooser.FILES_ONLY
    return if ((fileDialog.showSaveDialog(launcherFrame) == JFileChooser.CANCEL_OPTION)) null else fileDialog.selectedFile
}

fun createJButton(text: String, func: (e: ActionEvent) -> Unit) =
    JButton(text).apply {
        this.addActionListener { func(it) }
    }

/**
 * Get a empty label
 * */
fun emptyJLabel() = JLabel()
